src/Microsoft.AspNet.Mvc.Core/ModelBinding/CollectionModelBinder.cs

```diff
@@ -161,7 +188,18 @@ public class CollectionModelBinder<TElement> : IModelBinder
                 boundCollection.Add(ModelBindingHelper.CastOrDefault<TElement>(boundValue));
             }
 
-            return boundCollection;
+            return new CollectionResult
+            {
+                ValidationNode = validationNode,
+                Model = boundCollection
+            };
+        }
+
+        internal class CollectionResult
+        {
+            public ModelValidationNode ValidationNode { get; set; }
+
+            public IEnumerable<TElement> Model { get; set; }
```

Comment:
FYI Only: (previous API-review feedback) - the complications in this class really come from it have two very different code paths to do the same thing.

I feel like what's here is a good way to build this system based on the existing code, but a thorough refactor here would improve it.